mount/titus-mount-nfs.c

```diff
@@ -29,52 +29,66 @@ char nfs4[] = "nfs4";
 		}                                                              \
 	} while (0)
 
-static int dns_lookup(const char *hostname, struct sockaddr_in *addr)
-{
-	struct hostent *hp;
-	addr->sin_family = AF_INET;
+#define IP_ADDRESS_LEN 256
 
-	if (inet_aton(hostname, &addr->sin_addr)) {
-		fprintf(stderr,
-			"titus-mount-nfs: %s is already an IP. Not doing a DNS lookup\n",
-			hostname);
-		return 0;
-	}
-	fprintf(stderr,
-		"titus-mount-nfs: Decoding and resolving dns hostname for %s\n",
-		hostname);
-	hp = gethostbyname(hostname);
-	if (hp == NULL) {
-		int err_ret = h_errno;
-		fprintf(stderr, "titus-mount-nfs: can't get address for %s: %s\n",
-			hostname, hstrerror(err_ret));
-		return -1;
-	}
-	if (hp->h_length > (int)sizeof(struct in_addr)) {
-		fprintf(stderr, "titus-mount-nfs: got bad hp->h_length");
-		return -1;
-	}
-	memcpy(&addr->sin_addr, hp->h_addr, hp->h_length);
+static int add_addr_options(const char *hostname, char *final_options)
+{
+	strcat(final_options, ",addr=");
+	strcat(final_options, hostname);
 	return 0;
 }
 
-static void compose_final_options(const char *nfs_mount_hostname,
-				  char *final_options)
+/*
+ * Perform a DNS lookup and add to the DNS options. Prefer v6 over v4 
+ * (rely on /etc/gai.conf)
+ */
+static int hostname_to_ip_option(const char *hostname, char *final_options)
 {
-	/* For NFS, we must do the dns resolution *here* while we are inside net ns */
-	struct sockaddr_in server_addr;
-	char *ip_string;
+	struct addrinfo *result, *iter;
+	int error;
+	unsigned char buf[sizeof(struct in6_addr)];
 
-	if (dns_lookup(nfs_mount_hostname, &server_addr)) {
-		fprintf(stderr, "titus-mount-nfs: DNS lookup failed for %s",
-			nfs_mount_hostname);
-		exit(1);
+	error = inet_pton(AF_INET, hostname, buf);
+	if (error == 1) {
+		fprintf(stderr, "Valid IPv4 address %s - No DNS needed", hostname);
+		add_addr_options(hostname, final_options);
+		return 0;
+	} 
+
+	error = inet_pton(AF_INET6, hostname, buf);
+	if (error == 1) {
+		fprintf(stderr, "Valid IPv6 address %s - No DNS needed", hostname);
+		add_addr_options(hostname, final_options);
+		return 0;
+	} 
+
+	fprintf(stderr, "Now trying to do getaddrinfo %s: ", gai_strerror(error));
+	error = getaddrinfo(hostname, NULL, NULL, &result);
+	if (error) {
+		fprintf(stderr, "Error trying to resolve %s: ", gai_strerror(error));
```

Comment:
fprintf(stderr, "Error trying to resolve %s: %s", hostname, gai_strerror(error));

Reply:
Fixed.